client/notifyreport.cpp

```diff
@@ -9,135 +9,68 @@
  */
 
 #include "notifyreport.h"
-// Qt
-#include <QMouseEvent>
-#include <QPainter>
+
 // client
 #include "dialogs_g.h"
-// gui-qt
 #include "fc_client.h"
 #include "fonts.h"
 #include "page_game.h"
 #include "views/view_map.h"
 
+// Qt
+#include <QApplication>
+#include <QGridLayout>
+#include <QMouseEvent>
+
 /**
    Constructor for notify dialog
  */
-notify_dialog::notify_dialog(const char *caption, const char *headline,
-                             const char *lines, QWidget *parent)
-    : fcwidget(), label(nullptr), layout(nullptr)
+notify_dialog::notify_dialog(const QString &caption, const QString &headline,
+                             const QString &lines, QWidget *parent)
+    : fcwidget(), m_caption(caption), m_headline(headline)
 {
-  int x, y;
-  QString qlines;
-
   setAttribute(Qt::WA_DeleteOnClose);
+  setAttribute(Qt::WA_NoMousePropagation);
   setCursor(Qt::ArrowCursor);
   setParent(parent);
   setFrameStyle(QFrame::Box);
-  cw = new close_widget(this);
-  cw->put_to_corner();
-
-  qcaption = QString(caption);
-  qheadline = QString(headline);
-  qlines = QString(lines);
-  qlist = qlines.split(QStringLiteral("\n"));
-  small_font = fcFont::instance()->getFont(
-      QStringLiteral("gui_qt_font_notify_label"));
-  x = 0;
-  y = 0;
-  calc_size(x, y);
-  resize(x, y);
-  queen()->mapview_wdg->find_place(queen()->mapview_wdg->width() - x - 4, 4,
-                                   x, y, x, y, 0);
-  move(x, y);
-}
-
-/**
-   Starts new copy of notify dialog and closes current one
- */
-void notify_dialog::restart()
-{
-  QString s, q;
-  int i;
-  QByteArray capt_bytes;
-  QByteArray hl_bytes;
-  QByteArray qb_bytes;
-
-  for (i = 0; i < qlist.size(); ++i) {
-    s = qlist.at(i);
-    q = q + s;
-    if (i < qlist.size() - 1) {
-      q = q + QChar('\n');
-    }
-  }
-  capt_bytes = qcaption.toLocal8Bit();
-  hl_bytes = qheadline.toLocal8Bit();
-  qb_bytes = q.toLocal8Bit();
-  popup_notify_dialog(capt_bytes.data(), hl_bytes.data(), qb_bytes.data());
-  close();
-  destroy();
-}
 
-/**
-   Calculates size of notify dialog
- */
-void notify_dialog::calc_size(int &x, int &y)
-{
-  QFontMetrics fm(small_font);
-  int i;
-  QStringList str_list;
+  auto layout = new QGridLayout;
+  setLayout(layout);
 
-  str_list = qlist;
-  str_list << qcaption << qheadline;
+  m_contents =
+      new QLabel(QStringLiteral("%1 %2\n%3").arg(caption, headline, lines));
+  m_contents->setTextInteractionFlags(Qt::TextSelectableByMouse);
+  m_contents->setFont(fcFont::instance()->getFont(
+      QStringLiteral("gui_qt_font_notify_label")));
+  layout->addWidget(m_contents, 0, 0);
 
-  for (i = 0; i < str_list.count(); i++) {
-    x = qMax(x, fm.horizontalAdvance(str_list.at(i)));
-    y = y + 3 + fm.height();
-  }
-  y += fm.height();
-  x = x + 15;
-}
+  auto cw = new close_widget(this);
+  layout->addWidget(cw, 0, 1, Qt::AlignTop | Qt::AlignRight);
```

Comment:
In the other widgets with the `close_widget`, this method is used to place it `cw->put_to_corner();` that places the widget in the upper right corner. The way this is shown here is different. For consistency I think the close widget should be pushed to the corner.  Otherwise, looks great.